test/should_fail/fun_guard_info.erl

```diff
@@ -15,3 +17,20 @@ guard_unreachable_branch(Name)
     erlang:atom_to_list(Name);
 guard_unreachable_branch(_Name) ->
     [].
+
+% We know here that for sure at least one of them is an atom,
+% so this operation has no chance to pass.
+-spec at_least_one_of_them_is_an_atom(any(), any()) -> any().
+at_least_one_of_them_is_an_atom(A,B)
+  when is_atom(A), is_integer(B);
+       is_integer(A), is_atom(B) ->
+    A + B.
```

Comment:
What you're saying here is that we should treat `is_atom(A)` as `A :: atom()`. I'm not against this, but it's a new source of type information and we should consider what the implications are.

We have the following options, at least:

1. We accept these as type annotations
2. We don't use them, with the reasoning that code without specs shouldn't give type errors
3. We introduce an option.